### ISSUE TITLE
Latest -fsanitize=address findings

### ISSUE THREAD

**Issue by springmeyer** [mapnik/mapnik#3152]
With latest master seeing:

```
$ make test
Inheriting from mapnik-settings.env
* Starting Mapnik tests
* Running C++ Unit tests...


run is a Catch v1.1 b1 host application.
Run with -? for options

-------------------------------------------------------------------------------
geometry is_simple
  skipped is_simple tests
-------------------------------------------------------------------------------
test/unit/geometry/geometry_is_simple.cpp:8
...............................................................................

test/unit/geometry/geometry_is_simple.cpp:315: 
warning:
  geometry simple tests disabled due to boost version older that 1.58 used

=================================================================
==13401==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x619000306a80 at pc 0x00010e35e2c9 bp 0x7fff557134f0 sp 0x7fff55712cb0
READ of size 2048 at 0x619000306a80 thread T0
    #0 0x10e35e2c8 in __asan_memmove asan_interceptors.cc:425
    #1 0x10af7ae9b in mapnik::image<mapnik::rgba8_t>::image(mapnik::image<mapnik::rgba8_t> const&) (libmapnik.dylib+0x29ae9b)
    #2 0x10a7298f2 in ____C_A_T_C_H____T_E_S_T____10() (run+0x1002538f2)
    #3 0x10a9a370b in Catch::RunContext::invokeActiveTestCase() (run+0x1004cd70b)
    #4 0x10a9a160e in Catch::RunContext::runCurrentTest(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&) (run+0x1004cb60e)
    #5 0x10a996ee1 in Catch::RunContext::runTest(Catch::TestCase const&) (run+0x1004c0ee1)
    #6 0x10a986da6 in Catch::Runner::runTests() (run+0x1004b0da6)
    #7 0x10a97a417 in Catch::Session::run() (run+0x1004a4417)
    #8 0x10a95743a in main (run+0x10048143a)
    #9 0x7fff849755ac in start (libdyld.dylib+0x35ac)
    #10 0x0  (<unknown module>)

0x619000306a80 is located 0 bytes to the right of 1024-byte region [0x619000306680,0x619000306a80)
allocated by thread T0 here:
    #0 0x10e368c5b in wrap__Znwm asan_new_delete.cc:74
    #1 0x10af7a03b in mapnik::detail::buffer::buffer(unsigned long) (libmapnik.dylib+0x29a03b)
    #2 0x10a72772e in ____C_A_T_C_H____T_E_S_T____10() (run+0x10025172e)
    #3 0x10a9a370b in Catch::RunContext::invokeActiveTestCase() (run+0x1004cd70b)
    #4 0x10a9a160e in Catch::RunContext::runCurrentTest(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&) (run+0x1004cb60e)
    #5 0x10a996ee1 in Catch::RunContext::runTest(Catch::TestCase const&) (run+0x1004c0ee1)
    #6 0x10a986da6 in Catch::Runner::runTests() (run+0x1004b0da6)
    #7 0x10a97a417 in Catch::Session::run() (run+0x1004a4417)
    #8 0x10a95743a in main (run+0x10048143a)
    #9 0x7fff849755ac in start (libdyld.dylib+0x35ac)
    #10 0x0  (<unknown module>)

SUMMARY: AddressSanitizer: heap-buffer-overflow asan_interceptors.cc:425 in __asan_memmove
Shadow bytes around the buggy address:
  0x1c3200060d00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c3200060d10: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c3200060d20: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c3200060d30: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c3200060d40: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x1c3200060d50:[fa]fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c3200060d60: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c3200060d70: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c3200060d80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c3200060d90: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c3200060da0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==13401==ABORTING
-------------------------------------------------------------------------------
image class
  Image copy/move
-------------------------------------------------------------------------------
test/unit/imaging/image.cpp:10
...............................................................................

test/unit/imaging/image.cpp:293: FAILED:
  {Unknown expression after the reported line}
due to a fatal error condition:
  SIGABRT - Abort (abnormal termination) signal

===============================================================================
test cases:   22 |   21 passed | 1 failed
assertions: 5167 | 5166 passed | 1 failed
```

/cc @artemp for help triaging

**Comment by springmeyer** [comment]
Rebuilt with `DEBUG=True` and isolated to `test/unit/imaging/image.cpp`:

```
$ ./test/unit/run "image class"
=================================================================
==95058==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x61900001bc80 at pc 0x0001111822c9 bp 0x7fff5cd1f5d0 sp 0x7fff5cd1ed90
READ of size 2048 at 0x61900001bc80 thread T0
    #0 0x1111822c8 in __asan_memmove asan_interceptors.cc:425
    #1 0x10646b41f in mapnik::detail::buffer::buffer(mapnik::detail::buffer const&) algorithm:1766
    #2 0x10644546c in mapnik::detail::buffer::buffer(mapnik::detail::buffer const&) image.cpp:64
    #3 0x106445106 in mapnik::image<mapnik::rgba8_t>::image(mapnik::image<mapnik::rgba8_t> const&) image_impl.hpp:109
    #4 0x10644549c in mapnik::image<mapnik::rgba8_t>::image(mapnik::image<mapnik::rgba8_t> const&) image_impl.hpp:114
    #5 0x10353436f in ____C_A_T_C_H____T_E_S_T____10() image.cpp:345
    #6 0x103acdd43 in Catch::FreeFunctionTestCase::invoke() const catch.hpp:5818
    #7 0x103a9ad6b in Catch::TestCase::invoke() const catch.hpp:6723
    #8 0x103baa93c in Catch::RunContext::invokeActiveTestCase() catch.hpp:5417
    #9 0x103ba0df7 in Catch::RunContext::runCurrentTest(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&) catch.hpp:5389
    #10 0x103b7e58e in Catch::RunContext::runTest(Catch::TestCase const&) catch.hpp:5228
    #11 0x103b52476 in Catch::Runner::runTests() catch.hpp:5547
    #12 0x103b29091 in Catch::Session::run() catch.hpp:5678
    #13 0x103ac530f in Catch::Session::run(int, char* const*) catch.hpp:5658
    #14 0x103ac5188 in main run.cpp:12
    #15 0x7fff849755ac in start (libdyld.dylib+0x35ac)
    #16 0x1  (<unknown module>)

0x61900001bc80 is located 0 bytes to the right of 1024-byte region [0x61900001b880,0x61900001bc80)
allocated by thread T0 here:
    #0 0x11118cc5b in wrap__Znwm asan_new_delete.cc:74
    #1 0x10646ac50 in mapnik::detail::buffer::buffer(unsigned long) image.cpp:39
    #2 0x10644451c in mapnik::detail::buffer::buffer(unsigned long) image.cpp:41
    #3 0x103530d60 in ____C_A_T_C_H____T_E_S_T____10() image.cpp:295
    #4 0x103acdd43 in Catch::FreeFunctionTestCase::invoke() const catch.hpp:5818
    #5 0x103a9ad6b in Catch::TestCase::invoke() const catch.hpp:6723
    #6 0x103baa93c in Catch::RunContext::invokeActiveTestCase() catch.hpp:5417
    #7 0x103ba0df7 in Catch::RunContext::runCurrentTest(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&) catch.hpp:5389
    #8 0x103b7e58e in Catch::RunContext::runTest(Catch::TestCase const&) catch.hpp:5228
    #9 0x103b52476 in Catch::Runner::runTests() catch.hpp:5547
    #10 0x103b29091 in Catch::Session::run() catch.hpp:5678
    #11 0x103ac530f in Catch::Session::run(int, char* const*) catch.hpp:5658
    #12 0x103ac5188 in main run.cpp:12
    #13 0x7fff849755ac in start (libdyld.dylib+0x35ac)
    #14 0x1  (<unknown module>)

SUMMARY: AddressSanitizer: heap-buffer-overflow asan_interceptors.cc:425 in __asan_memmove
Shadow bytes around the buggy address:
  0x1c3200003740: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c3200003750: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c3200003760: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c3200003770: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c3200003780: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x1c3200003790:[fa]fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c32000037a0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c32000037b0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c32000037c0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c32000037d0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c32000037e0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==95058==ABORTING

run is a Catch v1.1 b1 host application.
Run with -? for options

-------------------------------------------------------------------------------
image class
  Image copy/move
-------------------------------------------------------------------------------
test/unit/imaging/image.cpp:10
...............................................................................

test/unit/imaging/image.cpp:293: FAILED:
  {Unknown expression after the reported line}
due to a fatal error condition:
  SIGABRT - Abort (abnormal termination) signal

test cases:    1 |    0 passed | 1 failed
assertions: 1179 | 1178 passed | 1 failed

```

**Comment by springmeyer** [comment]
this is happening on deep copy: https://github.com/mapnik/mapnik/blob/master/test/unit/imaging/image.cpp#L345

**Comment by artemp** [comment]
@springmeyer - doesn't look like a deep copy to me ^

**Comment by artemp** [comment]
```c++
// copy
buffer::buffer(buffer const& rhs)
    : size_(rhs.size_),
      data_(static_cast<unsigned char*>((rhs.owns_ && size_ != 0) ? ::operator new(size_) : nullptr)),
      owns_(rhs.owns_)
{
    if (data_) std::copy(rhs.data_, rhs.data_ + rhs.size_, data_);
    else data_ = rhs.data_;
}
```

I think the behaviour we're after is  ^ e.g only deep copy if rhs owns the data. 
/cc @springmeyer @flippmoke  - feedback ?

**Comment by artemp** [comment]
fixed in cb31692f3794259876b554ab96ac79a5b87508cf + 96c2fc9076e3a5273eba8f4951d5fa1661c780ea
